admin/fabfile.py

```diff
@@ -65,32 +66,146 @@ def bootstrap(ctx):
 
 @task
 def basicauth(ctx):
-    ctx.sudo('apt update')
-    ctx.sudo('apt install -y apache2-utils')
+    ctx.sudo("apt update")
+    ctx.sudo("apt install -y apache2-utils")
     # Will prompt for password.
-    ctx.sudo('htpasswd -c /etc/nginx/.htpasswd demozam')
+    ctx.sudo("htpasswd -c /etc/nginx/.htpasswd demozam")
 
 
 @task
 def letsencrypt(ctx):
-    ctx.sudo('add-apt-repository ppa:certbot/certbot')
-    ctx.sudo('apt update')
-    ctx.sudo('apt install -y certbot software-properties-common')
-    sudo_put(ctx, 'certbot.ini', '/srv/zam/certbot.ini')
-    sudo_put(ctx, 'ssl-renew', '/etc/cron.weekly/ssl-renew')
-    ctx.sudo('chmod +x /etc/cron.weekly/ssl-renew')
-    ctx.sudo('certbot certonly -c /srv/zam/certbot.ini --non-interactive '
-             '--agree-tos')
+    ctx.sudo("add-apt-repository ppa:certbot/certbot")
+    ctx.sudo("apt update")
+    ctx.sudo("apt install -y certbot software-properties-common")
+    sudo_put(ctx, "certbot.ini", "/srv/zam/certbot.ini")
+    sudo_put(ctx, "ssl-renew", "/etc/cron.weekly/ssl-renew")
+    ctx.sudo("chmod +x /etc/cron.weekly/ssl-renew")
+    ctx.sudo(
+        "certbot certonly -c /srv/zam/certbot.ini --non-interactive " "--agree-tos"
+    )
 
 
 @task
 def sshkeys(ctx):
-    for name, key in ctx.config.get('ssh_keys', {}).items():
-        ctx.run('grep -q -r "{key}" .ssh/authorized_keys '
-                '|| echo "{key}" '
-                '| sudo tee --append .ssh/authorized_keys'.format(key=key))
+    for name, key in ctx.config.get("ssh_keys", {}).items():
+        ctx.run(
+            'grep -q -r "{key}" .ssh/authorized_keys '
+            '|| echo "{key}" '
+            "| sudo tee --append .ssh/authorized_keys".format(key=key)
+        )
 
 
 @task
 def deploy(ctx, source):
-    put_dir(ctx, quote(source), '/srv/zam/', chown='zam')
+    put_dir(ctx, quote(source), "/srv/zam/", chown="zam")
+
+
+@task
+def deploy_repondeur(
+    ctx, secret, an_pattern_liste, an_pattern_amendement, branch="master"
+):
+    user = "repondeur"
+    create_user(ctx, name=user, home_dir="/srv/repondeur")
+    clone_repo(
+        ctx,
+        repo="https://github.com/betagouv/zam.git",
+        branch=branch,
+        path="/srv/repondeur/src",
+        user=user,
+    )
+    app_dir = "/srv/repondeur/src/repondeur"
+    install_requirements(ctx, app_dir=app_dir, user=user)
+    setup_config(ctx, app_dir=app_dir, user=user, secret=secret)
+    initialize_db(ctx, app_dir=app_dir, user=user)
+    fetch_an_group_data(ctx, user=user)
+    setup_service(ctx, an_pattern_liste, an_pattern_amendement)
+
+
+@task
+def create_user(ctx, name, home_dir):
+    if ctx.sudo(f"getent passwd {name}", warn=True, hide=True).failed:
```

Comment:
Maybe having a dedicated `bootstrap_repondeur` command would avoid most of extra checks?

Reply:
I understand that we may want to split between "things we only need to do once", and "things we need to do every time we deploy" for efficiency.

But then one day, to deploy the new version of the app, you need to add some new Debian package that the app now depends on. So you add it to the `bootstrap` task. But then to deploy that new version, you cannot just run `fab deploy`. You need to do `fab bootstrap` again, or it will fail.

So every time you deploy a new version, you need to think: what steps are needed this time? Do I need to run `bootstrap` again, or would `deploy` be enough?

Or you just run both `bootstrap` and `deploy` every time, just to be sure.

I don't like to think, so I kinda like having a "single command" idempotent deploy, to keep it simple and minimize errors, even if it's somewhat less efficient.